[CON-234] Enable write quorum

theoilie opened the pull request:
### Description

- Change `enforceWriteQuorum` env var default from false to true
- Make file consumers of the issueAndWaitForSecondarySyncRequests not await it because discovery only indexes metadata, so write quorum isn't needed on files
  - This applies to the routes: `/image_upload`, `/audius_users`, `/playlists`, `/tracks`
  - TODO: Can we confirm that data from these routes being unreplicated won't block indexing on discovery? Some of them were previously awaited but it seemed unnecessary since they didn't have metadata (but they do update the DB)
- Make metadata consumers of the issueAndWaitForSecondarySyncRequests middleware await it so that the middleware can actually block the request from succeeding until a 2/3 write quorum is reached
  - This makes writes await replication in all scenarios, but the write can still succeed if replication fails and write quorum is disabled
  - This applies to the routes: `/audius_users/metadata`, `/playlists/metadata`, `/tracks/metadata`
- Change order of precedence for enforcing client header vs server env var:
  - If client passes `Enforce-Write-Quorum` header as `false`, disable write quorum regardless of the `enforceWriteQuorum` env var
  - If client passes `Enforce-Write-Quorum` header as `true`, enable write quorum regardless of the `enforceWriteQuorum` env var
  - If client doesn't pass `Enforce-Write-Quorum` header, write quorum = `enforceWriteQuorum` env var



### Tests
Before each test: disable snapback (set disableSnapback env var to true) and refactored state monitoring (set stateMonitoringQueueRateLimitJobsPerInterval env var to 0) so that we can test only seeing manual syncs and not recurring
1. Default happy path with enforceWriteQuorum=false (now changed to default=true), default processSync behavior:
    * POST request to /audius_users/metadata when uploading track shows Enforce-Write-Quorum=undefined (still enqueued+awaited manual sync just didn't enforce failing the POST request if the sync failed)
    * Successfully logged completion before returning 200:
      * `issueAndWaitForSecondarySyncRequests - At least one secondary successfully replicated content for user 0xd7903b72f7e92f68ab14e61ef65689a201c4d741 in 534ms","time":"2022-07-08T00:20:17.617Z"
"requestMethod":"POST","requestHostname":"cn4_creator-node_1","requestUrl":"/audius_users/metadata","requestWallet":"0xd7903b72f7e92f68ab14e61ef65689a201c4d741","requestBlockchainUserId":"2","duration":627,"statusCode":200`
2. Expected failure case – enforceWriteQuorum=true, processSync temporarily changed to always throw:
    * Progress bar looks super glitchy – it bounced back and forth
    * Primary eventually returned 500 for the metadata POST request with the expected error:
      * `Error processing request: Error: issueAndWaitForSecondarySyncRequests Error - Failed to reach 2/3 write quorum for user`
      * This caused the UI to show the heavy load error. **TODO: Is this what we want? Should we try to issue a reconfig so they can write to a different RS?**
3. TODO: Enforce-Write-Quorum header = true, enforceWriteQuorum=false: should make the header take precedence (await and throw if 2/3 quorum isn't reached)
4. TODO: Enforce-Write-Quorum header = false, enforceWriteQuorum=true: should make the header take precedence (await but don't throw if 2/3 quorum isn't reached)
5. TODO: manualSyncsDisabled=true: should throw instead of return if enforceWriteQuorum=true
6. TODO: Missing req.session or req.session.wallet should throw instead of return if enforceWriteQuorum=true
7. TODO: Node is not primary or invalid creatorNodeEndpoints should throw instead of return if enforceWriteQuorum=true



### Monitoring - How will this change be monitored? Are there sufficient logs / alerts?
TODO: Search for success/failure logs and possibly add a metric

theoilie commented:
just a draft to make sure this is on the right track

theoilie commented on creator-node/src/middlewares.js:
```diff
@@ -266,18 +266,29 @@ async function ensureStorageMiddleware(req, res, next) {
 /**
  * Issue SyncRequests to both secondaries, and wait for at least one to sync before returning
  * @dev TODO - move out of middlewares layer
```
is there context on why this shouldn't be a middleware?

SidSethi replied:
yeah its not called as a middleware currently, just a function that some routes call

dmanjunath commented on creator-node/src/config.js:
```diff
@@ -497,13 +497,13 @@ const config = convict({
     format: Boolean,
     env: 'enforceWriteQuorum',
-    default: false
+    default: true
```
i think this is good to just have enabled and use the optimizely flag to control the rollout on prod. stage we would just 100% enable i'd think. so the steps could be 

1. define this optimizely flag and set it to false
2. turn it on for some % of users to observe behavior (and quickly turn off if we notice upload % drop significantly)
3. turn optimizely flag on for everyone
4. remove optimizely flag, default on for everyone

thoughts @SidSethi @theoilie/

SidSethi replied:
yeah that looks right

theoilie replied:
so staging can default to true for everyone, but would it make sense to default the env var to false on prod until enough clients actually send the env var? I just checked the client code and didn't see the env var actually defined anywhere so I guess I'd just add it [here](https://github.com/AudiusProject/audius-client/blob/30bbd598d5a9213358c2e4f0d0097c9b97ea2267/packages/web/src/common/services/remote-config/feature-flags.ts) and then pass it as the optimizely value in the routes mentioned in the PR?

dmanjunath commented on creator-node/src/middlewares.js:
```diff
@@ -266,18 +266,29 @@ async function ensureStorageMiddleware(req, res, next) {
     config.get('issueAndWaitForSecondarySyncRequestsPollingDurationMs')
+  // Write quorum header always takes precedence over env var if explicitly defined
+  // Empty/undefined header means enforceWriteQuorum decides if write quorum is enabled
+  const writeQuorumHeaderTrue = !!req.header('Enforce-Write-Quorum')
```
maybe it's just me but i found this logic a bit difficult to follow. i would suggest much more explicit cases that match your comment above like

```
const writeQuorumHeaderTrue = !!req.header('Enforce-Write-Quorum')
const writeQuorumHeaderFalse = req.header('Enforce-Write-Quorum') === false
let enforceWriteQuorum = false

if (!ignoreWriteQuorum) {
  if (writeQuorumHeaderTrue) enforceWriteQuorum = true
  // !writeQuorumHeaderFalse is for undefined/null/empty values where it's not explicitly false
  else if (!writeQuorumHeaderFalse && config.get('enforceWriteQuorum') enforceWriteQuorum = true
}
```

dmanjunath commented on creator-node/src/routes/audiusUsers.js:
```diff
@@ -176,7 +176,8 @@ module.exports = function (app) {
         await transaction.commit()
 
-        await issueAndWaitForSecondarySyncRequests(req)
+        // Discovery only indexes metadata and not files, so we eagerly replicate data but don't await it
```
thoughts on removing this non essential case entirely or commenting it out? thinking is we want these manual syncs to be as quick and uninterrupted as possible, cutting out all non essential jobs will help keep the queues more clear

SidSethi replied:
seems reasonable, only concern is added latency by not syncing early, bc now the final sync that is awaited has to cover a larger clock diff - in practice that would only really make a difference on heavy route like track upload so its probably fine? this is def worth adding a metric to measure duration of this particular sync request

good callout dheeraj

dmanjunath replied:
i actually thought about this more and we should keep this code in there. metadata syncs get issued immediately but if there's no other write after a route is called, the next time a sync could be triggered is recurring sync. we should definitely keep this in here

dmanjunath replied:
also i find it hilarious our await's were inverted based on what we needed

theoilie replied:
I see your reason for that and think it makes sense as long as we're okay with other data not being synced until a recurring job picks it up from state machine. alternatively, we could have another queue for these lower-priority manual syncs or add them to the recurring sync queue. which route do you think is preferable here?

dmanjunath replied:
yeah i think i was chatting with @SidSethi about this. i think your first instinct to keep the call without the await is the right one. the primary should make a request to the secondary to sync, doesn't need to block for a response

dmanjunath reviewed (commented):
first off thank you so much for testing this out and the great writeup. just had a few small suggestions but overall changes look good and i'm glad we're finally getting this out.

SidSethi commented:
thank you for putting in so much thought here and the thorough PR description 🙌

raymondjacobson commented on creator-node/src/middlewares.js:
```diff
@@ -266,18 +266,29 @@ async function ensureStorageMiddleware(req, res, next) {
     config.get('issueAndWaitForSecondarySyncRequestsPollingDurationMs')
+  // Write quorum header always takes precedence over env var if explicitly defined
+  // Empty/undefined header means enforceWriteQuorum decides if write quorum is enabled
+  const writeQuorumHeaderTrue = !!req.header('Enforce-Write-Quorum')
```
Just curious if the plan to have the client pass this header in as a feature flag?

theoilie replied:
yeah I think a feature flag makes the most sense for rollout like Dheeraj was saying above. I asked about the timeline of that in the discussion thread above as well

dmanjunath replied:
@raymondjacobson @theoilie yes that's the thinking. basically we can turn it on/off if upload rates degrade vs hardcoding as a server side env var which would take a release to flip. We actually have an optimizely flag but i don't think it ever got hooked up on the client/libs side

theoilie replied:
hmm I didn't see any flags when I searched "write" or "quorum" in optimizely but just made one this morning. I think I should prioritize the libs/client side of this first since that blocks rolling it out

raymondjacobson replied:
lmk if you have Qs on how to do that @theoilie

dmanjunath replied:
@theoilie here's the link https://app.optimizely.com/v2/projects/18140642080/features/21790311781/rules

dmanjunath replied:
@theoilie MVP

theoilie commented:
note that there's a bug in the axios version of libs that changes headers from `null` to `[object Object]`. this appears to be fixed in v0.20.0, but I'd rather not risk updating since I've read other users having issues after upgrading. the workaround I'm using is to pass the null as a string

see axios [solution PR](https://github.com/axios/axios/pull/2844) and problem [issue 1](https://github.com/axios/axios/issues/2701 ) and [issue 2](https://github.com/axios/axios/issues/2223)

theoilie commented:
@SidSethi or @dmanjunath  ready for review! I did some more testing, updated the PR description, and added a metric so we can track the rollout on staging and then prod. since there are only 7 routes that use write quorum, I think it's worth the slightly higher cardinality to have the `route` label that I added. If you're strongly opposed to this, though, I can change it to just be an `ignoreWriteQuorum` label. note that the `ignoreWriteQuorum` and `enforceWriteQuorum` labels should be the same for the same route on each request, so that lack of permutation should keep the cardinality down to a reasonable number

also I'll look into why the tests are failing tomorrow. I suspect it's because I changed the other routes to await write quorum

dmanjunath reviewed (approved):
This looks great to me, thanks for making these changes. I'm actually fine with the cardinality, don't see a problem with that at all, I love the observability 😃 . I'll let @SidSethi give the final approve but nothing else on my end

SidSethi commented on creator-node/src/middlewares.js:
```diff
@@ -247,32 +247,79 @@ async function ensureStorageMiddleware(req, res, next) {
 
```
ah also this would make an excellent demo lmao

theoilie commented on libs/src/services/creatorNode/CreatorNode.ts:
```diff
@@ -765,6 +769,11 @@ export class CreatorNode {
+      // Stringify `writeQuorumEnabled` to work around axios bug (fixed in axios v0.20.0): https://github.com/axios/axios/issues/2223
+      axiosRequestObj.headers['Enforce-Write-Quorum'] = String(
+        this.writeQuorumEnabled
+      )
```
I was trying to have some neutral default so the env var could control it if the client doesn't have a preference, but yeah it actually now defaults to 'false' anyway because the Optimizely flag only allows true/false. shouldn't be an issue either way now, although on the CN side we should support `undefined` since some clients won't send the param yet

SidSethi replied:
agreed on supporting this client side, just not sure it makes sense to default to non boolean on client side
don't think its an issue as is, just bit more complexity - your call

theoilie replied:
yeah I agree I'm not seeing any reason to default to null anymore. I'll change it and re-test before merging. I still have to fix all the tests that are failing anyway

dmanjunath commented:
@theoilie @SidSethi i think we can have libs code merged as part of this PR. we should just bump and merge the libs version separately

theoilie commented:
sounds good, just waiting to confirm that tests now pass on CI and then I'll merge!
